atecc608a_se.c

```diff
@@ -185,103 +185,100 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
         return PSA_ERROR_NOT_SUPPORTED;
     }
 
     /* The driver can only do randomized ECDSA on SHA-256 */
-    if (alg != PSA_ALG_ECDSA(PSA_ALG_SHA_256) && alg != PSA_ALG_ECDSA_ANY)
-    {
+    if (alg != PSA_ALG_ECDSA(PSA_ALG_SHA_256) && alg != PSA_ALG_ECDSA_ANY) {
         return PSA_ERROR_NOT_SUPPORTED;
     }
 
     ASSERT_SUCCESS_PSA(atecc608a_init());
 
-    ASSERT_SUCCESS(atcab_write_pubkey(key_id, pubkey_for_driver((uint8_t *) p_data)));
+    ASSERT_SUCCESS(atcab_write_pubkey(key_id, pubkey_for_driver((uint8_t *) data)));
+
+    if (bits != NULL) {
+        /* The 64-byte key is written as 72 bytes. See atcab_write_pubkey() for
+         * why 72 bytes. */
+        *bits = PSA_BYTES_TO_BITS(72);
+    }
+
 exit:
     atecc608a_deinit();
     return status;
 }
 
-static psa_status_t atecc608a_generate_key(psa_key_slot_number_t key_slot,
-                                           psa_key_type_t type,
-                                           psa_key_usage_t usage,
-                                           size_t bits,
-                                           const void *extra,
-                                           size_t extra_size,
-                                           uint8_t *p_pubkey_out,
-                                           size_t pubkey_out_size,
-                                           size_t *p_pubkey_length)
+static psa_status_t atecc608a_generate_key(
+    psa_drv_se_context_t *drv_context,
+    psa_key_slot_number_t key_slot,
+    const psa_key_attributes_t *attributes,
+    uint8_t *pubkey, size_t pubkey_size, size_t *pubkey_length)
 {
     const uint16_t key_id = key_slot;
     psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+    psa_key_type_t type = psa_get_key_type(attributes);
+    size_t bits = psa_get_key_bits(attributes);
 
     /* The hardware has slots 0-15 */
-    if (key_slot > 15)
-    {
+    if (key_slot > 15) {
         return PSA_ERROR_INVALID_ARGUMENT;
     }
 
-    if (type != PSA_KEY_TYPE_ECC_KEYPAIR(PSA_ECC_CURVE_SECP256R1))
-    {
+    if (type != PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_CURVE_SECP256R1)) {
         return PSA_ERROR_NOT_SUPPORTED;
     }
 
-    if (bits != PSA_BYTES_TO_BITS(ATCA_PRIV_KEY_SIZE))
-    {
+    if (bits != PSA_BYTES_TO_BITS(ATCA_PRIV_KEY_SIZE)) {
         return PSA_ERROR_NOT_SUPPORTED;
     }
 
-    if (p_pubkey_out != NULL && pubkey_out_size < 1 + ATCA_PUB_KEY_SIZE)
-    {
-       return PSA_ERROR_BUFFER_TOO_SMALL;
+    if (pubkey != NULL && pubkey_size < 1 + ATCA_PUB_KEY_SIZE) {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
     }
 
     ASSERT_SUCCESS_PSA(atecc608a_init());
 
-    if (p_pubkey_out != NULL)
-    {
-        ASSERT_SUCCESS(atcab_genkey(key_id, pubkey_for_driver(p_pubkey_out)));
-        pubkey_for_psa(p_pubkey_out);
-    }
-    else
-    {
+    if (pubkey != NULL) {
+        ASSERT_SUCCESS(atcab_genkey(key_id, pubkey_for_driver(pubkey)));
+        pubkey_for_psa(pubkey);
```

Comment:
Preexisting: `pubkey_for_driver` returns its result and has no side effect whereas `pubkey_for_psa` modifies its argument in place and returns void. The design of the functions make sense, but having similar naming makes it confusing. Please rename at least one of the functions.